File: movemork/Data/Repositories/MaintenanceRepository.swift
//
//  MaintenanceRepository.swift
//  movemork
//
//  MoveMark — Maintenance issue data access.
//  Attachment paths use `{userId}/{propertyId}/{issueId}/...` for `maintenance-media` RLS.
//

import Foundation
import Supabase

/// Fixed categories for maintenance issues; raw values match DB.
enum MaintenanceCategory: String, CaseIterable {
    case waterDamage = "Water damage"
    case heatingCooling = "Heating / cooling"
    case pest = "Pest"
    case appliance = "Appliance"
    case structural = "Structural"
    case plumbing = "Plumbing"
    case electrical = "Electrical"
    case other = "Other"
}

struct MaintenanceIssueRow: Codable, Identifiable {
    let id: UUID
    let propertyId: UUID
    let userId: UUID
    var title: String
    var description: String?
    var category: String?
    var status: String
    var dateDiscovered: String?
    var dateReported: String?
    var reportMethod: String?
    var landlordResponse: String?
    var followUpDate: String?
    var updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case propertyId = "property_id"
        case userId = "user_id"
        case title
        case description
        case category
        case status
        case dateDiscovered = "date_discovered"
        case dateReported = "date_reported"
        case reportMethod = "report_method"
        case landlordResponse = "landlord_response"
        case followUpDate = "follow_up_date"
        case updatedAt = "updated_at"
    }
}

struct MaintenanceRepository {

    func fetchIssues(propertyId: UUID) async throws -> [MaintenanceIssueRow] {
        try await supabase
            .from("maintenance_issues")
            .select()
            .eq("property_id", value: propertyId)
            .order("date_reported", ascending: false)
            .execute()
            .value
    }

    func insertIssue(_ row: MaintenanceIssueRow) async throws -> MaintenanceIssueRow {
        try await supabase
            .from("maintenance_issues")
            .insert(row)
            .select()
            .single()
            .execute()
            .value
    }

    func markResolved(id: UUID) async throws {
        let now = ISO8601DateFormatter().string(from: Date())
        try await supabase
            .from("maintenance_issues")
            .update(["status": "resolved", "updated_at": now])
            .eq("id", value: id)
            .execute()
    }

    func updateFollowUp(id: UUID, note: String) async throws {
        let now = ISO8601DateFormatter().string(from: Date())
        try await supabase
            .from("maintenance_issues")
            .update([
                "landlord_response": note,
                "follow_up_date": now,
                "updated_at": now
            ])
            .eq("id", value: id)
            .execute()
    }

    func uploadAttachment(data: Data, path: String) async throws -> String {
        try await supabase.storage
            .from("maintenance-media")
            .upload(path, data: data, options: FileOptions(contentType: "image/jpeg"))

        return path
    }
}

File: movemark-android/app/src/main/java/com/surensureshkumar/movemark/data/property/MaintenanceRepository.kt
package com.surensureshkumar.movemark.data.property

import com.surensureshkumar.movemark.data.models.MaintenanceIssueRow
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Order
import io.github.jan.supabase.storage.storage
import io.ktor.http.ContentType
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.time.Duration.Companion.seconds
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Serializable
private data class MaintenanceInsertRow(
    val id: String,
    @SerialName("property_id") val propertyId: String,
    @SerialName("user_id") val userId: String,
    val title: String,
    val category: String,
    val description: String,
    val status: String = "open",
)

@Serializable
private data class MaintenanceFollowUpUpdate(
    @SerialName("landlord_response") val landlordResponse: String,
    @SerialName("follow_up_date") val followUpDate: String,
)

@Serializable
private data class MaintenanceStatusUpdate(
    val status: String,
)

@Singleton
class MaintenanceRepository @Inject constructor(
    private val client: SupabaseClient,
) {
    companion object {
        const val MAINTENANCE_BUCKET = "maintenance-media"
    }

    suspend fun fetchIssues(propertyId: UUID): List<MaintenanceIssueRow> =
        client.from("maintenance_issues")
            .select {
                filter { eq("property_id", propertyId.toString()) }
                order("created_at", Order.DESCENDING)
            }
            .decodeList()

    suspend fun insertIssue(
        propertyId: UUID,
        userId: UUID,
        title: String,
        category: String,
        description: String,
    ): MaintenanceIssueRow {
        val id = UUID.randomUUID()
        val row = MaintenanceInsertRow(
            id = id.toString(),
            propertyId = propertyId.toString(),
            userId = userId.toString(),
            title = title.trim(),
            category = category.trim(),
            description = description.trim(),
        )
        return client.from("maintenance_issues")
            .insert(row) { select() }
            .decodeSingle()
    }

    suspend fun updateFollowUp(issueId: UUID, note: String) {
        val now = java.time.Instant.now().toString()
        client.from("maintenance_issues")
            .update(MaintenanceFollowUpUpdate(landlordResponse = note.trim(), followUpDate = now)) {
                filter { eq("id", issueId.toString()) }
            }
    }

    suspend fun markResolved(issueId: UUID) {
        client.from("maintenance_issues")
            .update(MaintenanceStatusUpdate(status = "resolved")) {
                filter { eq("id", issueId.toString()) }
            }
    }

    suspend fun uploadAttachment(data: ByteArray, path: String) {
        client.storage.from(MAINTENANCE_BUCKET).upload(path, data) {
            upsert = false
            contentType = ContentType.Image.JPEG
        }
    }

    suspend fun removeOrphanAttachment(path: String) {
        if (path.isEmpty()) return
        runCatching {
            client.storage.from(MAINTENANCE_BUCKET).delete(path)
        }
    }

    suspend fun signedUrl(path: String, expiresSeconds: Int = 3600): String =
        client.storage.from(MAINTENANCE_BUCKET).createSignedUrl(path, expiresSeconds.seconds)
}
